README.md
# a dead good html to rsx converter

A website to make it easier to convert html into rsx.

🦊


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file. See [conventional commits](https://www.conventionalcommits.org/) for commit guidelines.

- - -
## v0.4.5 - 2023-11-24
#### Bug Fixes
- update shuttle-secrets requirement from 0.33.0 to 0.34.0 - (37dc8c6) - dependabot[bot]
#### Continuous Integration
- bump cargo-bins/cargo-binstall from 1.4.5 to 1.4.6 - (c0a7aa8) - dependabot[bot]

- - -

## v0.4.4 - 2023-11-17
#### Bug Fixes
- update shuttle-secrets requirement from 0.32.0 to 0.33.0 - (35dcaf8) - dependabot[bot]
#### Continuous Integration
- bump cargo-bins/cargo-binstall from 1.4.4 to 1.4.5 - (195b8ae) - dependabot[bot]

- - -

## v0.4.3 - 2023-11-10
#### Bug Fixes
- update shuttle-secrets requirement from 0.31.0 to 0.32.0 - (67b8018) - dependabot[bot]

- - -

## v0.4.2 - 2023-11-04
#### Bug Fixes
- Support latest shuttle sdk - (9d56132) - Billie Thompson

- - -

## v0.4.1 - 2023-11-03
#### Bug Fixes
- update shuttle-secrets requirement from 0.30.1 to 0.31.0 - (21cadc4) - dependabot[bot]
#### Documentation
- Create LICENSE - (de5eafb) - Billie Thompson
- Create CODE_OF_CONDUCT.md - (564be7b) - Billie Thompson

- - -

## v0.4.0 - 2023-10-30
#### Bug Fixes
- Add instructions for crawlers and a name to the manifest - (d59c272) - Billie Thompson
#### Features
- Add compression to responses - (8aa884b) - Billie Thompson

- - -

## v0.3.1 - 2023-10-29
#### Bug Fixes
- Fix panics from new formatter - (ad4f489) - Billie Thompson

- - -

## v0.3.0 - 2023-10-29
#### Features
- Use the same parser as the cli - (9aaf636) - Billie Thompson
#### Refactoring
- Clippy advice - (3eb65d9) - Billie Thompson

- - -

## v0.2.4 - 2023-10-29
#### Bug Fixes
- Make the fox stay at the bottom better - (da1f39c) - Billie Thompson

- - -

## v0.2.3 - 2023-10-29
#### Bug Fixes
- update dead-good-html-to-rsx-converter requirement - (f91d75c) - dependabot[bot]
#### Documentation
- Add an important fox - (6d498ce) - Billie Thompson
- Remove the ugly homepage - (926e81c) - Billie Thompson

- - -

## v0.2.2 - 2023-10-29
#### Bug Fixes
- Make the page small enough not to trigger scroll bar - (c68abf5) - Billie Thompson

- - -

## v0.2.1 - 2023-10-29
#### Bug Fixes
- Correct the font used - (f867640) - Billie Thompson
#### Continuous Integration
- Add mergify config - (9417ab4) - Billie Thompson

- - -

## v0.2.0 - 2023-10-29
#### Continuous Integration
- Add api key - (2262327) - Billie Thompson
#### Features
- Put a cute little fox on the page - (5de61bd) - Billie Thompson

- - -

## v0.1.0 - 2023-10-29
#### Bug Fixes
- Linting - (2504470) - Billie Thompson
#### Continuous Integration
- Skip human interaction on binstall - (917727b) - Billie Thompson
- Correct binstall binary name - (b9955c8) - Billie Thompson
- Correct the skip human interaction check - (285060d) - Billie Thompson
- Skip human interaction on rustup - (b3bd49b) - Billie Thompson
- Add wasm to targets - (0eda18c) - Billie Thompson
- Add a deploy task - (60986b8) - Billie Thompson
- Ignore more files from the repo - (c9f2c3b) - Billie Thompson
- Add release configuration - (e0c1c21) - Billie Thompson
- Add ci - (24666cf) - Billie Thompson
#### Features
- Add support for deploying to shuttle.rs - (ac87cea) - Billie Thompson
- Upgrade favicon - (45e84a5) - Billie Thompson

- - -

Changelog generated by [cocogitto](https://github.com/cocogitto/cocogitto).

src/main.rs
#![allow(non_snake_case)]
#![allow(unused)]

use futures_util::stream::StreamExt;

use dead_good_html_to_rsx_converter::convert;
use dioxus::html::textarea;
use dioxus::prelude::*;
use dioxus_fullstack::prelude::*;
use dioxus_router::prelude::*;
use log::{warn, LevelFilter};
use serde::{Deserialize, Serialize};

fn main() {
    // Init debug
    dioxus_logger::init(LevelFilter::Info).expect("failed to init logger");

    let config = LaunchBuilder::<FullstackRouterConfig<Route>>::router();
    #[cfg(feature = "ssr")]
    let config = config.incremental(
        IncrementalRendererConfig::default().invalidate_after(std::time::Duration::from_secs(120)),
    );

    config.launch();
}

#[derive(Clone, Routable, Debug, PartialEq, Eq, Serialize, Deserialize)]
enum Route {
    #[route("/")]
    Home {},
}

#[inline_props]
fn Home(cx: Scope) -> Element {
    let parsed_text = use_state(cx, String::new);

    let text_parser = use_coroutine(cx, |mut rx| {
        to_owned![parsed_text];
        async move {
            loop {
                if let Some(text) = rx.next().await {
                    if let Ok(result) = convert(text) {
                        parsed_text.set(result);
                    }
                }
            }
        }
    });

    cx.render(rsx! {
        nav {
            class: "shadow-sm p-8 w-full bg-neutral-50 dark:bg-neutral-900 dark:text-white",
            h1 { "Dead good HTML to RSX Converter" }
        }
        div {
            class: "flex flex-col md:flex-row w-full items-stretch justify-stretch h-full p-4 gap-4",
            textarea {
                class: "resize-none px-0 w-full h-[85vh] text-sm text-neutral-900 bg-white border-0 dark:bg-neutral-800 focus:ring-0 dark:text-white dark:placeholder-neutral-400",
                placeholder: "Paste HTML here",
                required: true,
                oninput: move |evt| text_parser.send(evt.value.clone()),
            }
            textarea {
                class: "resize-none px-0 w-full h-[85vh] text-sm text-neutral-900 bg-white border-0 dark:bg-neutral-800 focus:ring-0 dark:text-white dark:placeholder-neutral-400",
                placeholder: "RSX will appear here",
                required: true,
                value: "{parsed_text}",
            }
        }
    })
}


src/lib.rs
#![allow(non_snake_case)]
#![allow(unused)]

use futures_util::stream::StreamExt;
use std::panic;

use dioxus::html::textarea;
use dioxus::prelude::*;
use dioxus_autofmt::write_block_out;
use dioxus_fullstack::prelude::*;
use dioxus_router::prelude::*;
use html_parser::Dom;
use log::{warn, LevelFilter};
use serde::{Deserialize, Serialize};

#[derive(Clone, Routable, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum AppRoutes {
    #[route("/")]
    Home {},
}

#[inline_props]
fn Home(cx: Scope) -> Element {
    let parsed_text = use_state(cx, String::new);

    let text_parser = use_coroutine(cx, |mut rx| {
        to_owned![parsed_text];
        async move {
            loop {
                if let Some(text) = rx.next().await {
                    let text: String = text;
                    if text.is_empty() {
                        // Shortcut if the text is empty
                        parsed_text.set(text);
                        continue;
                    }

                    let Ok(dom) = Dom::parse(text.trim()) else {
                        continue;
                    };

                    let body = rsx_rosetta::rsx_from_html(&dom);

                    let result = panic::catch_unwind(|| {
                        // This code has a lot of unwinds in it, so we need to catch them
                        write_block_out(body).expect("failed to write block out")
                    });

                    if let Ok(result) = result {
                        parsed_text.set(result);
                    }
                }
            }
        }
    });

    cx.render(rsx! {
        nav {
            class: "font-mono p-4 w-full bg-neutral-50 dark:bg-neutral-900 dark:text-white",
            h1 {
                class: "font-mono text-xl",
                "Dead good HTML to RSX converter"
            }
        }
        div {
            class: "font-mono flex flex-col flex-grow md:flex-row items-stretch justify-stretch p-4 gap-4",
            textarea {
                class: "resize-none p-4 w-full h-full text-sm text-neutral-900 bg-neutral-100 border-0 dark:bg-neutral-800 focus:ring-0 dark:text-white dark:placeholder-neutral-400",
                placeholder: "Paste HTML here",
                required: true,
                aria_label: "Paste HTML here",
                oninput: move |evt| text_parser.send(evt.value.clone()),
            }
            textarea {
                class: "font-mono resize-none p-4 w-full h-full text-sm text-neutral-900 bg-neutral-100 border-0 dark:bg-neutral-800 focus:ring-0 dark:text-white dark:placeholder-neutral-400",
                placeholder: "RSX will appear here",
                aria_label: "RSX will appear here",
                required: true,
                value: "{parsed_text}",
            }
        }
        div {
            class: "text-center font-mono p-4 w-full",
            "🦊"
        }
    })
}
